File: vicsek_simulation.jl
@inline function neighbors(d::Int)
    n = CartesianIndices((fill(-1:1, d)...,))
    return n[1:fld(length(n), 2)]
end

@inline function CellGrid(d::Int, L, r₀)
    indices = CartesianIndices((fill(-1:Int(fld(L,r₀)), d)...,))
    grid = Dict{CartesianIndex{d}, Vector{Int}}()
    for ci in indices
        grid[ci]=[]
    end
    return grid
end

function CellList(P, L, r₀, cells)
    n=length(P)
    d=length(P[1])
    for i in 1:n
        push!(cells[CartesianIndex(Int(fld(P[i][1],r₀)),Int(fld(P[i][2],r₀)))],i)
    end
    periodicBC(cells,L,r₀)
end

function ghostCells(L,r₀)
    tbBoundary=[CartesianIndex(i,Int(fld(L,r₀))-1) for i in 0:Int(fld(L,r₀))-1]
    append!(tbBoundary,[CartesianIndex(i,0) for i in 0:Int(fld(L,r₀))-1])
    rlBoundary=[CartesianIndex(Int(fld(L,r₀))-1,i) for i in 0:Int(fld(L,r₀))-1]
    append!(rlBoundary,[CartesianIndex(0,i) for i in 0:Int(fld(L,r₀))-1])
    return tbBoundary, rlBoundary
end

function periodicBC(cells,L,r₀)
    maxInd=Int(fld(L,r₀))
    tbBoundary, rlBoundary = ghostCells(L,r₀)
    for boundaryCell in tbBoundary
        if haskey(cells,boundaryCell)
            if boundaryCell[2]==0
                cells[boundaryCell+CartesianIndex(0,maxInd)] = cells[boundaryCell]
            elseif boundaryCell[2]==maxInd-1
                cells[boundaryCell+CartesianIndex(0,-maxInd)] = cells[boundaryCell]
            end
        end
    end
    for boundaryCell in rlBoundary
        if haskey(cells,boundaryCell)
            if boundaryCell[1]==0
                cells[boundaryCell+CartesianIndex(maxInd,0)] = cells[boundaryCell]
            elseif boundaryCell[1]==maxInd-1
                cells[boundaryCell+CartesianIndex(-maxInd,0)] = cells[boundaryCell]
            end
        end
    end
end

@inline function cellNeighbors!(ps, is, p, r)
    for (k, i) in enumerate(is[1:(end-1)])
        for j in is[(k+1):end]
            if norm(p[i]-p[j])<= r₀
                # println(i,j,p[i],p[j],norm(p[i]-p[j]))
                ps[i][j]=1
                ps[j][i]=1
            else
                ps[i][j]=0
                ps[j][i]=0
            end
        end
        ps[i][i]=1
    end
end

@inline function cellNeighbors!(ps, is, js, p, r)
    for i in is
        for j in js
            if norm(p[i]-p[j])<= r₀
                # println(i,j,p[i],p[j],norm(p[i]-p[j]))
                ps[i][j]=1
                ps[j][i]=1
            else
                ps[i][j]=0
                ps[j][i]=0
            end
        end
        ps[i][i]=1
    end
end

function near_neighbors(cells, P, r₀, FRN)
    offsets = neighbors(length(P[1]))
    # Iterate over non-empty cells
    for (cell, is) in cells
        # Pairs of points within the cell
        cellNeighbors!(FRN, is, P, r₀)
        # Pairs of points with non-empty neighboring cells
        for offset in offsets
            neigh_cell = cell + offset
            if haskey(cells, neigh_cell)
                @inbounds js = cells[neigh_cell]
                cellNeighbors!(FRN, is, js, P, r₀)
            end
        end
    end
end

function emptyCells(cells)
    for (ci,cell) in cells
        while length(cell)>0
            pop!(cell)
        end
    end
end

function fly(X,Θ,V,dt,L,r₀,N,v₀,η,cells,fixedRN)
    CellList(X, L, r₀, cells)
    # println("ok")
    near_neighbors(cells,X,r₀,fixedRN)
    # println("ok")
    
    Vmean=[normalize(mean(V[Is])) for Is in [findall(x->x>0,ff) for ff in fixedRN]]
    # println("ok")
    # Θnew=[u[2] > 0 ? acos(u[1])+η*(rand()-0.5) : 2π-acos(u[1])+η*(rand()-0.5) for u in Vmean]
    Θnew=[ atan(u[2],u[1])+η*(rand()-0.5) for u in Vmean]
    # println("ok")
    Vnew=[v₀*[cos(tt),sin(tt)] for tt in Θnew]
    # println("ok")
    
    for i in 1:N
        V[i]=Vnew[i]
        Θ[i]=Θnew[i]
        X[i]=X[i]+dt*Vnew[i]
    end
    # println("ok")
    PosPeriodicBC(X,L)
    # println("ok")
    
end

function PosPeriodicBC(X,L)
    for x in X
        x[1]=mod(x[1],L)
        x[2]=mod(x[2],L)
    end
end

function birdTrajectory(X,V,dt,N,nT,r₀,v₀,η,L,d)
    cells=CellGrid(d,L,r₀)
    fixedRN=[zeros(Int,N) for i in 1:N]
    filenameX=@sprintf("./vicsek_cells/positionsN%dL%2.lfr%2.lfv%2.lfeta%2.lf.csv",N,L,r₀,v₀,η)
    filenameV=@sprintf("./vicsek_cells/velocitiesN%dL%2.lfr%2.lfv%2.lfeta%2.lf.csv",N,L,r₀,v₀,η)
    Xio=open(filenameX,"w")
    Vio=open(filenameV,"w")
    writedlm(Xio, X, ", ")
    writedlm(Vio, V, ", ")
    close(Xio)
    close(Vio)
    Xio=open(filenameX,"a")
    Vio=open(filenameV,"a")
    for k in 1:nT
        fly(X,Θ,V,dt,L,r₀,N,v₀,η,cells,fixedRN)
        writedlm(Xio, X, ", ")
        writedlm(Vio, V, ", ")
        emptyCells(cells)
    end
    close(Xio)
    close(Vio)
end

function DataSet2(N,M,dt,nT,r₀,v₀,η,L,d)
    
    X₀=[[rand(2)*L for i in 1:N] for j in 1:M]
    Θ₀=[[2*π*rand() for i in 1:N] for j in 1:M]
    V₀=[[v₀*[cos(t),sin(t)] for t in tt] for tt in Θ₀]
    
    cells=CellGrid(d,L,r₀)
    fixedRN=[zeros(Int,N) for i in 1:N]
    for i in 1:N
        fixedRN[i][i]=1
    end
    
    for k in 1:nT
        fly(X₀[1],Θ₀[1],V₀[1],dt,L,r₀,N,v₀,η,cells,fixedRN)
        emptyCells(cells)
    end
    for i1 in 1:N
        for i2 in 1:N
            fixedRN[i1][i2]=0
        end
    end
    for i1 in 1:N
        fixedRN[i1][i1]=1
    end
    
    # S=[momentConfig(renormalisedData(X₀[1]),renormalisedData(V₀[1]))]
    S=[momentConfig(X₀[1] ./ L,V₀[1] ./ v₀)]
    
    for i in 2:M
        for k in 1:nT
            fly(X₀[i],Θ₀[i],V₀[i],dt,L,r₀,N,v₀,η,cells,fixedRN)
            emptyCells(cells)
        end
        
        for i1 in 1:N
        for i2 in 1:N
            fixedRN[i1][i2]=0
        end
        end
        for i1 in 1:N
            fixedRN[i1][i1]=1
        end
        # append!(S,[momentConfig(renormalisedData(X₀[i]),renormalisedData(V₀[i]))])
        append!(S,[momentConfig(X₀[i] ./ L,V₀[i] ./ v₀)])
    end

    return X₀, V₀, S
    
end

function dynamicDataset(N,M,dt,nT,r₀,v₀,η,L,d)
    X=Vector{Vector{Vector{Float64}}}(undef,0)
    X₀=[rand(2)*L for i in 1:N]
    Θ=Vector{Vector{Float64}}(undef,0)
    Θ₀=[2*π*rand() for i in 1:N]
    V=Vector{Vector{Vector{Float64}}}(undef,0)
    V₀=[v₀*[cos(t),sin(t)] for t in Θ₀]
    
    cells=CellGrid(d,L,r₀)
    fixedRN=[zeros(Int,N) for i in 1:N]
    for i in 1:N
        fixedRN[i][i]=1
    end
    
    push!(X,deepcopy(X₀))
    push!(Θ,deepcopy(Θ₀))
    push!(V,deepcopy(V₀))
    for k in 1:nT
        fly(X[end],Θ[end],V[end],dt,L,r₀,N,v₀,η,cells,fixedRN)
        emptyCells(cells)
    end
    for i1 in 1:N
        for i2 in 1:N
            fixedRN[i1][i2]=0
        end
    end
    for i1 in 1:N
        fixedRN[i1][i1]=1
    end
        
    S=[momentConfig((X[end] .- [[L/2,L/2]]) ./ L,V[end] ./ v₀)]
    
    for i in 2:M
        push!(X,deepcopy(X[end]))
        push!(Θ,deepcopy(Θ[end]))
        push!(V,deepcopy(V[end]))
        for k in 1:nT
            fly(X[end],Θ[end],V[end],dt,L,r₀,N,v₀,η,cells,fixedRN)
            emptyCells(cells)
        end
        for i1 in 1:N
            for i2 in 1:N
                fixedRN[i1][i2]=0
            end
        end
        for i1 in 1:N
            fixedRN[i1][i1]=1
        end
        append!(S,[momentConfig((X[end] .- [[L/2,L/2]]) ./ L,V[end] ./ v₀)])
        # S=[momentConfig(X[end] ./ L,V[end] ./ v₀)]
    end

    return X, V, S
    
end

function dynamicEnsemble(k,N,M,dt,nT,r₀,v₀,η,L,d)
    ensemble = [dynamicDataset(N,M,dt,nT,r₀,v₀,η,L,d)]
    for i in 2:k
        append!(ensemble,[dynamicDataset(N,M,dt,nT,r₀,v₀,η,L,d)])
    end
    return ensemble
end

#Minima for vicsek DataSet2 N=200;M=1000;dt=0.01;nT=200;r₀=2.0;v₀=3.0;η=0.05*π;L=10;d=2
#indices: 1083, 1432, 1473
#values: 0.0183, 0.082, 0.33

#Critical noise with N=200, L=10 is proportional to √(N/L²) = √2 ≃ 1.14
#But 0.5*π, π still orders
#I noticed that the simulation adds noise values between [-η/2,η/2] which means that the critical noise value we are going after is twice as large
#Therefore it makes sense that π still gives a phase transition

#η = 0.5π and the same values for the rest of the parameters is enough to generate a single trajectory that explores the entire slow manifold
# 1226
# 1482

function etaSet(N,M,dt,nT,r₀,v₀,ηs,L,d,ϵs,D;maxoutdim=3)
    
    filename1=@sprintf("./vicsekEvalsN%dL%2.lfr%.2lfv%.2lf.csv",N,L,r₀,v₀)
    Io1=open(filename1,"w")
    filename2=@sprintf("./vicsekOrderN%dL%2.lfr%.2lfv%.2lf.csv",N,L,r₀,v₀)
    Io2=open(filename2,"w")
    filename3=@sprintf("./vicsekSGEN%dL%2.lfr%.2lfv%.2lf.csv",N,L,r₀,v₀)
    Io3=open(filename3,"w")
       
    X,V,S=DataSet2(N,M,dt,nT,r₀,v₀,ηs[1],L,d)
    for i in eachindex(S)
        for j in 1:i-1
            D[i,j] = norm(S[i]-S[j])
            D[j,i] = D[i,j]
        end
    end
    SGE, ϵOpt = ϵOptimum(D,ϵs)
    transition_matrix, e_vals, e_vecs, modes = DiffMap(D, ϵOpt, maxoutdim = 3)
    writedlm(Io1, e_vals, ", ")
    close(Io1)
    writedlm(Io2, mean([sqrt(modes[2,i]^2+modes[3,i]^2) for i in 1:M]), ", ")
    close(Io2)
    writedlm(Io3, SGE, ", ")
    writedlm(Io3, ϵOpt)
    close(Io3)
    Io1=open(filename1,"a")
    Io2=open(filename2,"a")
    Io3=open(filename3,"a")
        
    for j in eachindex(ηs[2:end])
        X,V,S=DataSet2(N,M,dt,nT,r₀,v₀,ηs[j],L,d)
        for i in eachindex(S)
            for j in 1:i-1
                D[i,j] = norm(S[i]-S[j])
                D[j,i] = D[i,j]
            end
        end
        SGE, ϵOpt = ϵOptimum(D,ϵs)
        transition_matrix, e_vals, e_vecs, modes = DiffMap(D, ϵOpt, maxoutdim = maxoutdim)
        writedlm(Io1, e_vals, ", ")
        writedlm(Io2, mean([sqrt(modes[2,i]^2+modes[3,i]^2) for i in 1:M]), ", ")
        writedlm(Io3, SGE, ", ")
        writedlm(Io3, ϵOpt)
    end
    close(Io1)
    close(Io2)
    close(Io3)
end

function ϵOptimum(D,ϵs;ϵ_0=0.25)
    SGE = SGECriteria(D,ϵs)
    #Using the Peaks Library
    minima = findminima(SGE)
    if length(minima.indices) > 0
        return SGE, ϵs[minima.indices[end]];
    else
        return SGE, ϵ_0;
    end
end

function etaSet2(N,M,dt,nT,r₀,v₀,ηs,L,d,ϵs,ϵOpt,D;maxoutdim=3)
    
    filename1=@sprintf("./vicsekEvalsN%dL%2.lfr%.2lfv%.2lf.csv",N,L,r₀,v₀)
    Io1=open(filename1,"w")
    filename2=@sprintf("./vicsekOrderN%dL%2.lfr%.2lfv%.2lf.csv",N,L,r₀,v₀)
    Io2=open(filename2,"w")
       
    X,V,S=DataSet2(N,M,dt,nT,r₀,v₀,ηs[1],L,d)
    for i in eachindex(S)
        for j in 1:i-1
            D[i,j] = norm(S[i]-S[j])
            D[j,i] = D[i,j]
        end
    end
    
    transition_matrix, e_vals, e_vecs, modes = DiffMap(D, ϵOpt, maxoutdim = 3)
    writedlm(Io1, e_vals, ", ")
    close(Io1)
    writedlm(Io2, mean([sqrt(modes[2,i]^2+modes[3,i]^2) for i in 1:M]), ", ")
    close(Io2)
    Io1=open(filename1,"a")
    Io2=open(filename2,"a")
        
    for j in eachindex(ηs[2:end])
        X,V,S=DataSet2(N,M,dt,nT,r₀,v₀,ηs[j],L,d)
        for i in eachindex(S)
            for j in 1:i-1
                D[i,j] = norm(S[i]-S[j])
                D[j,i] = D[i,j]
            end
        end
        transition_matrix, e_vals, e_vecs, modes = DiffMap(D, ϵOpt, maxoutdim = maxoutdim)
        writedlm(Io1, e_vals, ", ")
        writedlm(Io2, mean([sqrt(modes[2,i]^2+modes[3,i]^2) for i in 1:M]), ", ")
    end
    close(Io1)
    close(Io2)
end

File: vicsek_simulation.py
#Vicsek simulation using intrinsic noise in reorientation updating rule and nearest neighbor simulation using a cell list.
#Probably there is an easier and more direct way to do this in python

import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import find_peaks

# Calculates the nearest neighbors cells in a d-dimensional space.
# Returns a list of tuples representing the neighboring cells.
def neighbors(d: int):
    indices = np.ndindex((3,) * d)  # Create indices for all neighbors including self
    return list(indices)[: len(indices) // 2]  # Return only the first half to avoid double counting

# Partitions the space into cells of size r_0 in a d-dimensional space with side length L.
# Returns a dictionary where keys are tuples representing cell positions and
# values are empty lists to store particle indices.
def CellGrid(d: int, L, r_0):
    indices = np.ndindex((int(L // r_0) + 1,) * d)  # Create indices for all cells including ghost cells
    grid = {index: [] for index in indices}  # Initialize a dictionary to store the grid
    return grid

# Creates a cell list for particles P in a d-dimensional space with side length L and interaction radius r_0.
# Assigns each particle to its corresponding cell in the grid `cells`.
# Includes ghost cells to handle periodic boundary conditions.
def CellList(P, L, r_0, cells):
    n = len(P)  # Number of particles
    for i in range(n):
        # Assign particle i to its cell based on its position
        cell_index = tuple(int(P[i][j] // r_0) for j in range(len(P[i])))
        cells[cell_index].append(i)
    periodicBC(cells, L, r_0)  # Apply periodic boundary conditions to the cell list

# Returns the tuples representing the ghost cells for a d-dimensional space with side length L and cell size r_0.
# Ghost cells are copies of boundary cells used to handle periodic boundary conditions.
def ghostCells(L, r_0):
    max_index = int(L // r_0)
    tbBoundary = [(i, max_index) for i in range(max_index + 1)]  # Top boundary cells
    tbBoundary += [(i, 0) for i in range(max_index + 1)]  # Bottom boundary cells
    rlBoundary = [(max_index, i) for i in range(max_index + 1)]  # Right boundary cells
    rlBoundary += [(0, i) for i in range(max_index + 1)]  # Left boundary cells
    return tbBoundary, rlBoundary

# Applies periodic boundary conditions to the cell list `cells`.
# Copies the particles in boundary cells to their corresponding ghost cells.
def periodicBC(cells, L, r_0):
    max_index = int(L // r_0)
    tbBoundary, rlBoundary = ghostCells(L, r_0)  # Get the ghost cell indices

    # Handle top and bottom boundary conditions
    for boundaryCell in tbBoundary:
        if boundaryCell in cells:  # Check if the boundary cell exists
            if boundaryCell[1] == 0:  # If it's a bottom boundary cell
                cells[(boundaryCell[0], max_index)] = cells[boundaryCell]  # Copy to top ghost cell
            elif boundaryCell[1] == max_index:  # If it's a top boundary cell
                cells[(boundaryCell[0], 0)] = cells[boundaryCell]  # Copy to bottom ghost cell

    # Handle right and left boundary conditions
    for boundaryCell in rlBoundary:
        if boundaryCell in cells:  # Check if the boundary cell exists
            if boundaryCell[0] == 0:  # If it's a left boundary cell
                cells[(max_index, boundaryCell[1])] = cells[boundaryCell]  # Copy to right ghost cell
            elif boundaryCell[0] == max_index:  # If it's a right boundary cell
                cells[(0, boundaryCell[1])] = cells[boundaryCell]  # Copy to left ghost cell

# Calculates the nearest neighbors within a cell or between two cells.
# Updates the `ps` matrix (fixed radius neighbors matrix) based on distances between particles.
def cellNeighbors(ps, Is, p, r_0):
    for k, i in enumerate(Is[:-1]):  # Iterate over all pairs of particles in the cell
        for j in Is[k + 1:]:
            if np.linalg.norm(np.array(p[i]) - np.array(p[j])) <= r_0:  # Check if the distance is within the interaction radius
                ps[i][j] = 1  # Mark as neighbors
                ps[j][i] = 1
            else:
                ps[i][j] = 0
                ps[j][i] = 0
        ps[i][i] = 1  # Each particle is a neighbor of itself

# Calculates the nearest neighbors between two cells.
# Updates the `ps` matrix (fixed radius neighbors matrix) based on distances between particles.
def cellNeighbors2(ps, Is, js, p, r_0):
    for i in Is:  # Iterate over particles in the first cell
        for j in js:  # Iterate over particles in the second cell
            if np.linalg.norm(np.array(p[i]) - np.array(p[j])) <= r_0:  # Check if the distance is within the interaction radius
                ps[i][j] = 1  # Mark as neighbors
                ps[j][i] = 1
            else:
                ps[i][j] = 0
                ps[j][i] = 0
        ps[i][i] = 1  # Each particle is a neighbor of itself

# Computes the nearest neighbors matrix `FRN` for particles P using a cell list.
# FRN_{ij} = 1 if the distance between p_i and p_j is less than the interaction radius r_0.
def near_neighbors(cells, P, r_0, FRN):
    offsets = neighbors(len(P[0]))  # Get the neighboring cell offsets

    # Iterate over non-empty cells
    for cell, Is in cells.items():
        # Pairs of points within the cell
        cellNeighbors(FRN, Is, P, r_0)  # Calculate neighbors within the cell

        # Pairs of points with non-empty neighboring cells
        for offset in offsets:
            neigh_cell = tuple(np.array(cell) + np.array(offset))  # Get the neighboring cell
            if neigh_cell in cells:  # Check if the neighboring cell exists
                js = cells[neigh_cell]  # Get the particles in the neighboring cell
                cellNeighbors2(FRN, Is, js, P, r_0)  # Calculate neighbors between the cells

# Empties all cells in the cell list `cells`.
def emptyCells(cells):
    for cell in cells.values():
        cell.clear()  # Clear the list of particles in each cell

# Evolves the Vicsek model for one time step.
# Updates the positions (X), orientations (Theta), and velocities (V) of the particles.
def fly(X, Theta, V, dt, L, r_0, N, v_0, eta, cells, fixedRN):
    CellList(X, L, r_0, cells)  # Create the cell list
    near_neighbors(cells, X, r_0, fixedRN)  # Compute the nearest neighbors matrix

    # Calculate the mean velocity of neighbors for each particle
    Vmean = [
        np.mean(np.array([V[j] for j in np.where(fixedRN[i] > 0)[0]]), axis=0)
        for i in range(N)
    ]
    Vmean = [v / np.linalg.norm(v) if np.linalg.norm(v) > 0 else np.array([1, 0]) for v in Vmean]  # Normalize

    # Update the orientations with intrinsic noise
    Theta_new = [np.arctan2(u[1], u[0]) + eta * (np.random.rand() - 0.5) for u in Vmean]

    # Update the velocities based on the new orientations
    V_new = np.array([[v_0 * np.cos(tt), v_0 * np.sin(tt)] for tt in Theta_new])

    # Update the positions, orientations, and velocities of the particles
    for i in range(N):
        V[i] = V_new[i]
        Theta[i] = Theta_new[i]
        X[i] = X[i] + dt * V_new[i]

    PosPeriodicBC(X, L)  # Apply periodic boundary conditions to the positions

# Applies periodic boundary conditions to the positions of the particles.
def PosPeriodicBC(X, L):
    for i in range(len(X)):
        X[i][0] = X[i][0] % L  # Apply periodic boundary condition in the x-dimension
        X[i][1] = X[i][1] % L  # Apply periodic boundary condition in the y-dimension

# Simulate one trajectory
def birdTrajectory(X, V, dt, N, nT, r_0, v_0, eta, L, d):
    cells = CellGrid(d, L, r_0)  # Create the cell grid
    fixedRN = np.zeros((N, N), dtype=int)  # Initialize the nearest neighbors matrix

    # Create filenames for saving positions and velocities
    filenameX = f"./vicsek_cells/positionsN{N}L{L:.1f}r{r_0:.1f}v{v_0:.1f}eta{eta:.2f}.csv"
    filenameV = f"./vicsek_cells/velocitiesN{N}L{L:.1f}r{r_0:.1f}v{v_0:.1f}eta{eta:.2f}.csv"

    # Prepare data for saving
    X_data = np.array(X)
    V_data = np.array(V)

    # Save initial positions and velocities
    np.savetxt(filenameX, X_data, delimiter=",")
    np.savetxt(filenameV, V_data, delimiter=",")

    for k in range(nT):
        fly(X, Theta, V, dt, L, r_0, N, v_0, eta, cells, fixedRN)  # Evolve the system for one time step

        # Append current positions and velocities to data arrays
        X_data = np.array(X)
        V_data = np.array(V)

        # Append data to files
        with open(filenameX, "a") as fX, open(filenameV, "a") as fV:
            np.savetxt(fX, X_data, delimiter=",")
            np.savetxt(fV, V_data, delimiter=",")

        emptyCells(cells)  # Empty the cells for the next time step


# Generates a dataset of M realizations of the Vicsek model with N particles and nT time steps.
# Returns only the last state of each realization.
def DataSet2(N, M, dt, nT, r_0, v_0, eta, L, d):
    X_0 = np.random.rand(M, N, 2) * L  # Initialize random positions
    Theta_0 = np.random.rand(M, N) * 2 * np.pi  # Initialize random orientations
    V_0 = np.array([[[v_0 * np.cos(t), v_0 * np.sin(t)] for t in tt] for tt in Theta_0])  # Initialize velocities

    cells = CellGrid(d, L, r_0)  # Create the cell grid
    fixedRN = np.zeros((N, N), dtype=int)  # Initialize the nearest neighbors matrix
    np.fill_diagonal(fixedRN, 1)  # Each particle is a neighbor of itself

    # Evolve the first realization for nT time steps
    for _ in range(nT):
        fly(X_0[0], Theta_0[0], V_0[0], dt, L, r_0, N, v_0, eta, cells, fixedRN)
        emptyCells(cells)

    S = [momentConfig(X_0[0] / L, V_0[0] / v_0)]  # Calculate the moment configuration for the first realization

    # Evolve the remaining realizations for nT time steps
    for i in range(1, M):
        for _ in range(nT):
            fly(X_0[i], Theta_0[i], V_0[i], dt, L, r_0, N, v_0, eta, cells, fixedRN)
            emptyCells(cells)
        S.append(momentConfig(X_0[i] / L, V_0[i] / v_0))  # Calculate the moment configuration

    return X_0, V_0, S

# Returns a trajectory of length nT*M, sampled at every nT time steps
def dynamicDataset(N, M, dt, nT, r_0, v_0, eta, L, d):
    X = []  # Initialize an empty list to store positions
    X_0 = np.random.rand(N, 2) * L  # Initialize random positions
    Theta = []  # Initialize an empty list to store orientations
    Theta_0 = np.random.rand(N) * 2 * np.pi  # Initialize random orientations
    V = []  # Initialize an empty list to store velocities
    V_0 = np.array([[v_0 * np.cos(t), v_0 * np.sin(t)] for t in Theta_0])  # Initialize velocities

    cells = CellGrid(d, L, r_0)  # Create the cell grid
    fixedRN = np.zeros((N, N), dtype=int)  # Initialize the nearest neighbors matrix
    np.fill_diagonal(fixedRN, 1)  # Each particle is a neighbor of itself

    X.append(X_0.copy())  # Add the initial positions to the dataset
    Theta.append(Theta_0.copy())  # Add the initial orientations to the dataset
    V.append(V_0.copy())  # Add the initial velocities to the dataset

    # Evolve the system for nT time steps
    for _ in range(nT):
        fly(X[-1], Theta[-1], V[-1], dt, L, r_0, N, v_0, eta, cells, fixedRN)
        emptyCells(cells)

    S = [momentConfig(X_0 / L, V_0 / v_0)]  # Calculate the moment configuration for the initial state

    # Evolve the system for nT time steps for each realization
    for _ in range(1, M):
        X.append(X[-1].copy())  # Add the previous positions to the dataset
        Theta.append(Theta[-1].copy())  # Add the previous orientations to the dataset
        V.append(V[-1].copy())  # Add the previous velocities to the dataset
        for _ in range(nT):
            fly(X[-1], Theta[-1], V[-1], dt, L, r_0, N, v_0, eta, cells, fixedRN)
            emptyCells(cells)
        S.append(momentConfig(X[-1] / L, V[-1] / v_0))  # Calculate the moment configuration

    return X, V, S

def dynamicEnsemble(k, N, M, dt, nT, r_0, v_0, eta, L, d):
    ensemble = [dynamicDataset(N, M, dt, nT, r_0, v_0, eta, L, d)]
    for _ in range(1, k):
        ensemble.append(dynamicDataset(N, M, dt, nT, r_0, v_0, eta, L, d))
    return ensemble

# Calculated the optimum value of epsilon for a given number of particles, speed,
# interaction radius and noise intensity
def etaSet(N, M, dt, nT, r_0, v_0, etas, L, d, epsilons, D, maxoutdim=3):
    # Create filenames for saving data
    filename1 = f"./vicsekEvalsN{N}L{L:.1f}r{r_0:.2f}v{v_0:.2f}.csv"
    filename2 = f"./vicsekOrderN{N}L{L:.1f}r{r_0:.2f}v{v_0:.2f}.csv"
    filename3 = f"./vicsekSGEN{N}L{L:.1f}r{r_0:.2f}v{v_0:.2f}.csv"

    # Calculate and save data for each eta value
    for eta in etas:
        X, V, S = DataSet2(N, M, dt, nT, r_0, v_0, eta, L, d)
        for i in range(len(S)):
            for j in range(i):
                D[i, j] = np.linalg.norm(S[i] - S[j])
                D[j, i] = D[i, j]

        SGE, epsilonOpt = epsilonOptimum(D, epsilons)
        _, e_vals, _, _ = DiffMap(D, epsilonOpt, maxoutdim=maxoutdim)

        # Append data to files
        with open(filename1, "a") as f1, open(filename2, "a") as f2, open(filename3, "a") as f3:
            np.savetxt(f1, e_vals, delimiter=",")
            np.savetxt(f2, np.mean([np.sqrt(modes[1, i] ** 2 + modes[2, i] ** 2) for i in range(M)]), delimiter=",")
            np.savetxt(f3, SGE, delimiter=",")
            np.savetxt(f3, [epsilonOpt], delimiter=",")

# Calculates local minima, returns the largest value of epsilon that attains a minimum
# if it exists otherwise a fixed value
def epsilonOptimum(D, epsilons, epsilon_0=0.25):
    SGE = SGECriteria(D, epsilons)

    # Use scipy.signal.find_peaks to find local minima
    minima_indices, _ = find_peaks(-SGE)  # Find peaks of the inverted SGE
    if len(minima_indices) > 0:
        return SGE, epsilons[minima_indices[-1]]
    else:
        return SGE, epsilon_0